pkg/loghttp/push/otlp.go

```diff
@@ -0,0 +1,373 @@
+	mostRecentEntryTimestamp time.Time
+	contentType              string
+	contentEncoding          string
+	bodySize                 int64
+}
+
+func newPushStats() *Stats {
+	return &Stats{
+		logLinesBytes:           map[time.Duration]int64{},
+		structuredMetadataBytes: map[time.Duration]int64{},
+	}
+}
+
+func ParseOTLPRequest(userID string, r *http.Request, tenantsRetention TenantsRetention) (*logproto.PushRequest, *Stats, error) {
+	stats := newPushStats()
+	otlpLogs, err := extractLogs(r, stats)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	req := otlpToLokiPushRequest(otlpLogs, userID, tenantsRetention, stats)
+	return req, stats, nil
+}
+
+func extractLogs(r *http.Request, pushStats *Stats) (plog.Logs, error) {
+	pushStats.contentEncoding = r.Header.Get(contentEnc)
+	// bodySize should always reflect the compressed size of the request body
+	bodySize := loki_util.NewSizeReader(r.Body)
+	var body io.Reader = bodySize
+	if pushStats.contentEncoding == gzipContentEncoding {
+		r, err := gzip.NewReader(bodySize)
+		if err != nil {
+			return plog.NewLogs(), err
+		}
+		body = r
+		defer func(reader *gzip.Reader) {
+			_ = reader.Close()
+		}(r)
+	}
+	buf, err := io.ReadAll(body)
+	if err != nil {
+		return plog.NewLogs(), err
+	}
+
+	pushStats.bodySize = bodySize.Size()
+
+	req := plogotlp.NewExportRequest()
+
+	pushStats.contentType = r.Header.Get(contentType)
+	switch pushStats.contentType {
+	case pbContentType:
+		err := req.UnmarshalProto(buf)
+		if err != nil {
+			return plog.NewLogs(), err
+		}
+	case applicationJSON:
+		err := req.UnmarshalJSON(buf)
+		if err != nil {
+			return plog.NewLogs(), err
+		}
+	default:
+		return plog.NewLogs(),
+			errors.Errorf(
+				"content type: %s is not supported",
+				r.Header.Get("Content-Type"),
+			)
+	}
+
+	return req.Logs(), nil
+}
+
+func otlpToLokiPushRequest(ld plog.Logs, userID string, tenantsRetention TenantsRetention, stats *Stats) *logproto.PushRequest {
+	if ld.LogRecordCount() == 0 {
+		return &logproto.PushRequest{}
+	}
+
+	rls := ld.ResourceLogs()
+	pushRequestsByStream := make(map[string]logproto.Stream, rls.Len())
+
+	for i := 0; i < rls.Len(); i++ {
+		sls := rls.At(i).ScopeLogs()
+		res := rls.At(i).Resource()
+
+		flattenedResourceAttributes := labels.NewBuilder(logproto.FromLabelAdaptersToLabels(attributesToLabels(res.Attributes(), "")))
+		// service.name is a required Resource Attribute. If it is not present, we will set it to "unknown_service".
+		if flattenedResourceAttributes.Get("service_name") == "" {
+			flattenedResourceAttributes = flattenedResourceAttributes.Set("service_name", "unknown_service")
+		}
+
+		if dac := res.DroppedAttributesCount(); dac != 0 {
+			flattenedResourceAttributes = flattenedResourceAttributes.Set("resource_dropped_attributes_count", fmt.Sprintf("%d", dac))
+		}
+
+		// copy blessed attributes to stream labels
+		streamLabels := make(model.LabelSet, len(blessedAttributesNormalized))
+		for _, ba := range blessedAttributesNormalized {
```

Comment:
The blessed attributes should be an ordered list by preference.

Reply:
We are not going to enforce `max_label_names_per_series` limit here. The limits would be enforced later at the common code path which processed the parsed push requests. It would be the user's responsibility to set appropriate blessed attributes(we are going to make it configurable) to avoid rejecting data when the label count goes beyond the limits.